godot-core/src/obj/oneditor.rs

```diff
@@ -0,0 +1,307 @@
+///        }
+///     }
+/// }
+///```
+///
+/// ## Example - factory pattern
+///
+/// ```
+/// use godot::prelude::*;
+///
+/// #[derive(GodotClass)]
+/// #[class(init, base = Node)]
+/// struct SomeClass {
+///     #[export]
+///     required_node: OnEditor<Gd<Node>>,
+/// }
+///
+/// fn create_and_add(
+///     mut this: Gd<Node>,
+///     some_class_scene: Gd<PackedScene>,
+///     some_node: Gd<Node>,
+/// ) -> Gd<SomeClass> {
+///     let mut my_node = some_class_scene.instantiate_as::<SomeClass>();
+///
+///     // Would cause the panic:
+///     // this.add_child(&my_node);
+///
+///     // Note: Remember that nodes are manually managed.
+///     // They will leak memory if not added to tree and/or pruned.
+///     my_node.bind_mut().required_node.init(some_node);
+///
+///     // Will not cause the panic.
+///     this.add_child(&my_node);
+///
+///     my_node
+/// }
+/// ```
+///
+/// # Using `OnEditor<T>` with other GodotTypes
+///
+/// `OnEditor<T>` can be used with other built-ins to provide extra validation logic and making sure that given properties has been set.
+/// Example usage might be checking if entities has been granted properly generated id.
+///
+/// In such cases the value which will be deemed invalid **must** be specified with `#[init(uninit = val)]`.
+/// Given `val` will be used to represent uninitialized `OnEditor<T>` in the Godot editor.
+/// Accessing uninitialized value will cause the panic.
+///
+/// ## Example - using `OnEditor` with primitives
+///
+/// ```
+///  use godot::prelude::*;
+///
+/// #[derive(GodotClass)]
+/// #[class(init, base = Node)]
+/// struct SomeClassThatCanBeInstantiatedInCode {
+///     // Uninitialized value will be represented by `42` in the editor.
+///     // Will cause panic if not set via the editor or code before use.
+///     #[export]
+///     #[init(invalid = 42)]
+///     some_primitive: OnEditor<i64>,
+/// }
+///
+/// fn create_and_add(mut this: Gd<Node>, val: i64) -> Gd<SomeClassThatCanBeInstantiatedInCode> {
+///     let mut my_node = SomeClassThatCanBeInstantiatedInCode::new_alloc();
+///
+///     // Would cause the panic:
+///     // this.add_child(&my_node);
+///
+///     my_node.bind_mut().some_primitive.init(val);
+///
+///     // Will not cause the panic.
+///     this.add_child(&my_node);
+///
+///     my_node
+/// }
+/// ```
+///
+/// # Using `OnEditor<T>` with `#[class(tool)]`
+///
+/// When used with `#[class(tool)]`, the before-ready checks are omitted.
+/// Otherwise, `OnEditor<T>` behaves the same — accessing an uninitialized value
+/// will cause a panic.
+pub struct OnEditor<T> {
+    inner: OnEditorState<T>,
+}
+
+pub(crate) enum OnEditorState<T> {
+    /// Uninitialized null value.
+    Null,
+    /// Uninitialized state, but with a value marked as invalid (required to represent non-nullable type in the editor).
+    Uninitialized(T),
+    /// Initialized with a value.
+    Initialized(T),
+}
+
+impl<T: Var + FromGodot + PartialEq> OnEditor<T> {
```

Comment:
Maybe add brief comment that `PartialEq` is needed to compare against uninitialized sentinel values.

Reply:
By the way, `Sentiel` might be good name for `invalid` :thinking:

Reply:
Yep, I mentioned it here :slightly_smiling_face: 
https://github.com/godot-rust/gdext/pull/1051#pullrequestreview-2634783549

Will still think about those names...